### ISSUE TITLE
Batch query can be very slow for values() step even with small limit

### ISSUE THREAD

**Issue by li-boxuan** [JanusGraph/janusgraph#2592]
I found this problem on 0.5.3 with `query.batch = true`. I believe this problem exists on master as well.

Consider a graph with many (say, millions of nodes). Consider below queries:

1. `g.V().values().limit(1).profile()`
2. `g.V().limit(1).values().profile()`
3. `g.V().valueMap().limit(1).profile()`

The 2rd and 3rd queries run very fast, which is expected. The 1st one, however, is very slow, and I suspect it actually scans the entire graph and fetches all vertices. 

I believe this is consistent with what @rngcntr explained in the opening comment of https://github.com/JanusGraph/janusgraph/pull/2516 - quoted here:

> The reason is that with MultiQuery enabled, each VertexStep becomes a blocking operator, which aggregates all traversers from the previous operator before firing a query.

Apparently, JanusGraph attempts to aggregates all vertices `g.V()` before proceeding with `values().limit(1)`. This is very bad because we only need one vertex.

In general, for a traversal ending with `values().limit(x)`, the limit `x` should be propagated to the "blocking operator`.

**Comment by li-boxuan** [comment]
Note that this is different from `g.V(0).out().out().limit(1)`. The first `out()` might return many neighbors but we don't know whether these neighbors have `out()`, so it is arguably legitimate to aggregate the results of first `out()` before proceeding with second `out()`.

For case `g.V().values().limit(1)`, we know for sure we only need one vertex, so there is no justification to aggregate all `V()` before proceeding with `values()`.

**Comment by rngcntr** [comment]
A simple optimization strategy should work well here. All we need to do is move `limit(n)` steps after `values()` steps in front of the `values()` step, right?

**Comment by li-boxuan** [comment]
> A simple optimization strategy should work well here. All we need to do is move limit(n) steps after values() steps in front of the values() step, right?

Sounds like a plan. An alternative is to check whether the next step is limit step, in `JanusGraphPropertiesStep`. Not sure if that is possible and which option is better.

**Comment by rngcntr** [comment]
In my opinion, a strategy would be better for two reasons:

- Although being slightly less efficient (have to scan the entire query for `limit()` steps), rearranging the steps before execution will allow the users to inspect what is actually executed using `explain()`.
- I'm pretty sure this is not the only case where a limit can be applied earlier that it currently is. In such an optimization strategy, we could collect a number of steps, which can benefit of an early limit and have it all in one place.

**Comment by FlorianHockmann** [comment]
There is already a strategy in TinkerPop that should move all `limit()` steps as far as possible to the left: [EarlyLimitStrategy](https://github.com/apache/tinkerpop/blob/master/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/EarlyLimitStrategy.java). It was created for the ticket [TINKERPOP-1882](https://issues.apache.org/jira/browse/TINKERPOP-1882) which reads nearly like a duplicate of this issue here.

I just checked the first traversal `g.V().values().limit(1).profile()` here and the problem is simply that the limit is not applied here to limit the number of vertices but the number of property values. If it would limit the number of vertices, then the strategy could move the `limit()` step to the left. But now it has to evaluate `values()` first and only then apply the `limit()`. It works however as expected if `valueMap()` is used instead of `values()`.

Here are the profiles for both traversals:

```
gremlin> g.V().values().limit(1).profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphStep(vertex,[])                                                                     50.981    99.35
    \_condition=()
    \_orders=[]
    \_isFitted=false
    \_isOrdered=true
    \_query=[]
  optimization                                                                                 2.713
  optimization                                                                                 0.011
  scan                                                                                         0.000
    \_query=[]
    \_fullscan=true
    \_condition=VERTEX
JanusGraphPropertiesStep(value)                                                                0.098     0.19
RangeGlobalStep(0,1)                                                                           0.237     0.46
                                            >TOTAL                     -           -          51.317        -
```

```
gremlin> g.V().valueMap().limit(1).profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphStep(vertex,[])                                                                      0.736    96.02
    \_condition=()
    \_orders=[]
    \_limit=1
    \_isFitted=false
    \_isOrdered=true
    \_query=[]
  optimization                                                                                 0.013
  optimization                                                                                 0.006
  scan                                                                                         0.000
    \_query=[]
    \_fullscan=true
    \_condition=VERTEX
PropertyMapStep(value)                                                                         0.030     3.98
                                            >TOTAL                     -           -           0.766        -
```

**Comment by rngcntr** [comment]
I somehow thought, a `values()` step without a parameter would return a key-value map, but instead this step is a flat map step. In this case, it does indeed not make sense to pull `limit()` before `values()` as that would change the query semantically.

**Comment by li-boxuan** [comment]
> as that would change the query semantically.

@rngcntr Would you mind explaining why that changes the semantics? I thought they are the same.

**Comment by rngcntr** [comment]
@li-boxuan If we have multiple properties on each vertex, then the behavior would be different:
```
g.V().limit(2).values()
==> v0_prop0_val
==> v0_prop1_val
==> v1_prop0_val
==> v1_prop1_val
```
```
g.V().values().limit(2)
==> v0_prop0_val
==> v0_prop1_val
```

**Comment by li-boxuan** [comment]
@rngcntr You are absolutely correct; we usually only use values("propKey") thus forgot about that difference.

Then I guess the optimization could be:

For a values step which contains a single property key, which is of SINGLE cardinality (thus, only one value), we could move the limit step before values() step.

**Comment by rngcntr** [comment]
But wouldn't that still cause differences if the property is set for some but not all vertices? `values("prop0").limit(1)` would search vertices until it finds one that has this property, whereas `limit(1).values("prop0")` would take only one vertex and then return either one or no property value.

**Comment by li-boxuan** [comment]
> values("prop0").limit(1) would search vertices until it finds one that has this property, whereas limit(1).values("prop0") would take only one vertex and then return either one or no property value.

Interesting... You are right, this seems to be the behaviour. Do you know if valueMap() should behave similarly or not? Currently valueMap("prop0").limit(1) returns nothing if the first vertex does not contain this property.

**Comment by rngcntr** [comment]
My naive understanding is that `valueMap()` on a property that does not exist should simply return an empty map. Since an empty map is still a result, it satisfies the condition of `limit(1)` and therefore the query terminates, returning `[]`. I just verified that and I consider that expected behavior.

**Comment by li-boxuan** [comment]
@rngcntr You are right, that is also consistent with TinkerGraph's behavior:
```
gremlin> graph = TinkerGraph.open()
==>tinkergraph[vertices:0 edges:0]
gremlin> g = graph.traversal()
==>graphtraversalsource[tinkergraph[vertices:0 edges:0], standard]
gremlin> g.addV().property("prop0","val0")
==>v[0]
gremlin> g.addV().property("prop1","val1")
==>v[2]
gremlin> g.V().values("prop1").limit(1)
==>val1
gremlin> g.V().valueMap("prop1").limit(1)
==>[]
```
